native/dlaln2.go

```diff
@@ -0,0 +1,386 @@
+//
+// If both singular values of (ca A - w D) are less than smin, smin*identity
+// will be used instead of (ca A - w D). If only one singular value is less than
+// smin, one element of (ca A - w D) will be perturbed enough to make the
+// smallest singular value roughly smin. If both singular values are at least
+// smin, (ca A - w D) will not be perturbed. In any case, the perturbation will
+// be at most some small multiple of max(smin, ulp*norm(ca A - w D)). The
+// singular values are computed by infinity-norm approximations, and thus will
+// only be correct to a factor of 2 or so.
+//
+// All input quantities are assumed to be smaller than overflow by a reasonable
+// factor.
+//
+// scale is a scaling factor less than or equal to 1 which is chosen so that X
+// can be computed without overflow. X is further scaled if necessary to assure
+// that norm(ca A - w D)*norm(X) is less than overflow.
+//
+// xnorm contains the infinity-norm of X when X is regarded as a na×nw real
+// matrix.
+//
+// ok will be false if (ca A - w D) had to be perturbed to make its smallest
+// singular value greater than smin, otherwise ok will be true.
+//
+// Dlaln2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlaln2(trans bool, na, nw int, smin, ca float64, a []float64, lda int, d1, d2 float64, b []float64, ldb int, wr, wi float64, x []float64, ldx int) (scale, xnorm float64, ok bool) {
+	if na != 1 && na != 2 {
+		panic("lapack: invalid value of na")
+	}
+	if nw != 1 && nw != 2 {
+		panic("lapack: invalid value of nw")
+	}
+	checkMatrix(na, na, a, lda)
+	checkMatrix(na, nw, b, ldb)
+	checkMatrix(na, nw, x, ldx)
+
+	smlnum := 2 * dlamchS
+	bignum := 1 / smlnum
+	smini := math.Max(smin, smlnum)
+
+	ok = true
+	scale = 1
+
+	if na == 1 {
+		// 1×1 (i.e., scalar) system C X = B.
+
+		if nw == 1 {
+			// Real 1×1 system.
+
+			// C = ca A - w D.
+			csr := ca*a[0] - wr*d1
+			cnorm := math.Abs(csr)
+
+			// If |C| < smini, use C = smini.
+			if cnorm < smini {
+				csr = smini
+				cnorm = smini
+				ok = false
+			}
+
+			// Check scaling for X = B / C.
+			bnorm := math.Abs(b[0])
+			if cnorm < 1 && bnorm > math.Max(1, bignum*cnorm) {
+				scale = 1 / bnorm
+			}
+
+			// Compute X.
+			x[0] = b[0] * scale / csr
+			xnorm = math.Abs(x[0])
+
+			return scale, xnorm, ok
+		}
+
+		// Complex 1×1 system (w is complex).
+
+		// C = ca A - w D.
+		csr := ca*a[0] - wr*d1
+		csi := -wi * d1
+		cnorm := math.Abs(csr) + math.Abs(csi)
+
+		// If |C| < smini, use C = smini.
+		if cnorm < smini {
+			csr = smini
+			csi = 0
+			cnorm = smini
+			ok = false
+		}
+
+		// Check scaling for X = B / C.
+		bnorm := math.Abs(b[0]) + math.Abs(b[1])
+		if cnorm < 1 && bnorm > math.Max(1, bignum*cnorm) {
+			scale = 1 / bnorm
+		}
+
+		// Compute X.
+		cx := complex(scale*b[0], scale*b[1]) / complex(csr, csi)
+		x[0], x[1] = real(cx), imag(cx)
+		xnorm = math.Abs(x[0]) + math.Abs(x[1])
```

Comment:
I agree that's what Lapack does, but that's a strange definition of complex norm...